pkg/statistics/histogram.go

```diff
@@ -1528,10 +1522,85 @@ func MergePartitionHist2GlobalHist(sc *stmtctx.StatementContext, hists []*Histog
 				sum += buckets[i-1].Count
 				bucketNDV += buckets[i-1].NDV
 			}
-			merged, err := mergePartitionBuckets(sc, buckets[i:r])
+			res, err := currentLeftMost.Compare(sc.TypeCtx(), buckets[i].lower, collate.GetBinaryCollator())
+			if err != nil {
+				return nil, err
+			}
+			if res > 0 {
+				currentLeftMost = buckets[i].lower
+			}
+
+			// Iterate possible overlapped ones.
+			// We need to re-sort this part.
+			mergeBuffer = mergeBuffer[:0]
+			fullyInsideBuffer = fullyInsideBuffer[:0]
+			cutAndFixBuffer = cutAndFixBuffer[:0]
+			origI := i
+			for ; i > 0; i-- {
+				res, err := buckets[i-1].upper.Compare(sc.TypeCtx(), currentLeftMost, collate.GetBinaryCollator())
+				if err != nil {
+					return nil, err
+				}
+				// If buckets[i-1].upper <= currentLeftMost, this bucket has no overlap with current merging one. Break it.
+				if res <= 0 {
+					break
+				}
+				// Now the bucket[i-1].upper > currentLeftMost, they are overlapped.
+				res, err = buckets[i-1].lower.Compare(sc.TypeCtx(), currentLeftMost, collate.GetBinaryCollator())
+				if err != nil {
+					return nil, err
+				}
+				// If buckets[i-1].lower >= currentLeftMost, this bucket is totally inside. So it can be totoally merged.
+				if res >= 0 {
+					sum += buckets[i-1].Count
+					bucketNDV += buckets[i-1].NDV
+					fullyInsideBuffer = append(fullyInsideBuffer, buckets[i-1])
+					mergeBuffer = append(mergeBuffer, buckets[i])
+					continue
+				}
+				// Now buckets[i-1].lower < currentLeftMost < buckets[i-1].upper
+				overlapping := calcFraction4Datums(buckets[i-1].lower, buckets[i-1].upper, currentLeftMost)
+				overlappedCount := int64(float64(buckets[i-1].Count) * overlapping)
+				overlappedNDV := int64(float64(buckets[i-1].NDV) * overlapping)
+				sum += overlappedCount
+				bucketNDV += overlappedNDV
+				buckets[i-1].Count -= overlappedCount
+				buckets[i-1].NDV -= overlappedNDV
+
+				// Cut it.
+				cutBkt := newBucket4Meging()
+				buckets[i-1].upper.Copy(cutBkt.upper)
+				currentLeftMost.Copy(cutBkt.lower)
+				cutBkt.Count = overlappedCount
+				cutBkt.NDV = overlappedNDV
+				mergeBuffer = append(mergeBuffer, cutBkt)
+				cutAndFixBuffer = append(cutAndFixBuffer, cutBkt)
+
+				currentLeftMost.Copy(buckets[i-1].upper)
+				buckets[i-1].Repeat = 0
```

Comment:
Don't we need a `break` for the last case here?

Reply:
No, there'll be multiple buckets that meet the same requirements. `[small, currentLeftMost+1]`, `[small2, currentLeftMost+2]`, `[small3, currentLeftMost+x]`...